### DISCUSSION TITLE
After adding to nextjs+prisma app custom Prisma Client extension I get following error "PrismaClient is not configured to run in Vercel Edge Functions or Edge Middleware"

### DISCUSSION THREAD

**Discussion by berezhinskiyo** [prisma/prisma#23724] [Q&A]
### Question

I have added to nextjs+prisma app my custom Prisma Client extension, and I got following error "PrismaClient is not configured to run in Vercel Edge Functions or Edge Middleware". 
<img width="509" alt="image" src="https://github.com/prisma/prisma/assets/20126398/822eeb2d-e0bb-4a4d-ac6a-b63eb66327cd">
Developers of nextjs+prisma app told that they didn't use any edge function, but they use middleware.
When I remove extension - everything works well.
 How to check edge functions in app, why it involves to PrismaClient extension? Give me some advice please.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @berezhinskiyo 👋 

The error message you're seeing is related to the fact that Prisma Client is not designed to run in Vercel Edge Functions or Edge Middleware. This is because these environments do not allow communication with databases via the normal methods like TCP connections, which Prisma usually uses. See the comment in this [issue](https://github.com/prisma/prisma/issues/21310#issuecomment-1960350838).

Next.js middlewares are always run in an edge runtime, not in Node. Unfortunately that means you can only use the edge functions support of Prisma and not falls back to the classic built-in drivers - even locally. See this Next.js documentation: https://nextjs.org/docs/app/building-your-application/routing/middleware#runtime

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by berezhinskiyo** [reply]
Thank you, @ludralph. But why it works without extension definition for PrismaClient?
`const prisma = new PrismaClient()` - in this case is doesn't rise an exception

**Reply by berezhinskiyo** [reply]
May be you mean using Prisma from middleware functions, but I don't do it, I use PrismaClient only in regular functions for CRUD.

**Reply by ludralph** [reply]
Hi @berezhinskiyo 

It's possible that when you add a custom Prisma Client extension, it seems to trigger this compatibility issue with the edge runtime.
When you use a plain `PrismaClient()` initialization, Prisma might not immediately detect that it's running in an edge environment, especially if no database operations are performed right away.

However, when you add an extension to PrismaClient, it triggers additional checks and initializations that are more likely to detect the edge runtime environment. This detection then leads to the error you're seeing.
This behavior is consistent with what's described in the [Github Issue about Edge Runtime support](https://github.com/prisma/prisma/issues/20560), where it's mentioned that `Prisma Client` fails to initialize in edge environments.
It's important to note that even though `const prisma = new PrismaClient()` doesn't immediately raise an exception, you might still encounter issues when trying to perform database operations in an edge environment.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
